### PR TITLE
Fix Topology Manager policy and scope not being updated after NRT creation

### PR DESCRIPTION
**Pull request by PiotrProkop**
fixes https://github.com/kubernetes-sigs/node-feature-discovery/issues/1254

### CODE REVIEW

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:309]: Ideally, I'd prefer having a goroutine waiting for changes in the kubeletconfig and re-reading it if needed and only if needed, avoding constant and costly polling. IIRC we use already `fsnotify` in NFD so this would mean no new deps. Have you explored this option? how hard would it be to implement?

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:309]: I feel like if we want to support both endpoints and not to overcomplicate the code. We can spin up another goroutine that will check `kubelet` config every 5mins(or make it configurable) to reduce costly poling and `patch` only `TopologyPolices` or attributes field?

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:309]: Fair point. I'm hesitant to add kubelet config read in this path, this is why I mentioned a different goroutine.
Let me review the code again and think a bit more.

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:309]: I feel like spinning up another goroutine which will be updating NRT object can lead to conflicts when updating same resource it would be best to have one func that's responsible for managing NRT. We can also consider such implementation:
```
-       policy, scope, err := w.detectTopologyPolicyAndScope()
-       if err != nil {
-               return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
-       }
-
-       tmAttributes := createTopologyAttributes(policy, scope)
-       deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
-
        nrt, err := cli.TopologyV1alpha2().NodeResourceTopologies().Get(context.TODO(), w.nodeName, metav1.GetOptions{})
        if errors.IsNotFound(err) {
+               policy, scope, err := w.detectTopologyPolicyAndScope()
+               if err != nil {
+                       return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
+               }
+
+               tmAttributes := createTopologyAttributes(policy, scope)
+               deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
+
                nrtNew := v1alpha2.NodeResourceTopology{
                        ObjectMeta: metav1.ObjectMeta{
                                Name: w.nodeName,
@@ -237,15 +239,28 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
        nrtMutated := nrt.DeepCopy()
        nrtMutated.Zones = zoneInfo

-       attributes := append(scanResponse.Attributes, tmAttributes...)
-       updateAttributes(&nrtMutated.Attributes, attributes)
+       if w.lastUpdateTimestamp.IsZero() || time.Since(w.lastUpdateTimestamp).Minutes() > float64(5) {
+               policy, scope, err := w.detectTopologyPolicyAndScope()
+               if err != nil {
+                       return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
+               }

-       nrtMutated.TopologyPolicies = deprecatedTopologyPolicies
+               tmAttributes := createTopologyAttributes(policy, scope)
+               deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
+
+               attributes := append(scanResponse.Attributes, tmAttributes...)
+               updateAttributes(&nrtMutated.Attributes, attributes)
+
+               nrtMutated.TopologyPolicies = deprecatedTopologyPolicies
+       }

        nrtUpdated, err := cli.TopologyV1alpha2().NodeResourceTopologies().Update(context.TODO(), nrtMutated, metav1.UpdateOptions{})
        if err != nil {
                return fmt.Errorf("failed to update NodeResourceTopology: %w", err)
        }
+
+       w.lastUpdateTimestamp = time.Now()
+
```

or move reading `scope` and `policy` into seperate goroutine and keep them in shared state somewhere?

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:326]: @ffromani to reduce number of times we need to read Kubelet config file and minimize complexity. I have modified the code only to reread Kubelet config when `Interval based` event is triggered, which default to 1 min(somewhat low). Should we introduce another ticker? 
I feel like we should implement a better solution to this error but I am unsure if we need to do it in this PR.

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:326]: thanks @PiotrProkop my main issue was finding time to review, I'll have another pass ASAP. I tend to agree without a some redesign (which is better served on a separate PR) we can't do much better than the current approach.

**Review by marquiz — COMMENTED** [review]
Thanks @PiotrProkop for the enhancement. I have just a few notes after a quick read-through, relying mostly on @ffromani here :)

One ask: you have nice description in the related github issue, could you take some of that to the commit message?

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:207]: 👍

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:253]: I'll try to think of something.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:178]: Checking now.

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:178]: If I understand correctly current behavior there is no first run during startup. The first update happens either after `sleepInterval` or when nfd detects changes to checkpoint files.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: I think this is cleaner now 👍

**Comment by marquiz** [comment]
Thx for the updates @PiotrProkop. I don't think I have much to add. I'll leave this to @ffromani now and will approve when he's happy

**Comment by marquiz** [comment]
ping @ffromani

**Comment by ffromani** [comment]
> ping @ffromani

Hi! sorry for the delay I'll review by the end of this week (July 28)

**Comment by marquiz** [comment]
> Hi! sorry for the delay I'll review by the end of this week (July 28)

Thanks @ffromani, no panic with this (yet), just wanted to remind so that this wouldn't be the last one pending

**Review by marquiz — APPROVED** [review]
Thanks @ffromani. I agree that going forward some re-architecting would be beneficial, e.g. detecting kubelet restarts or smth

**Comment by PiotrProkop** [comment]
```
 /home/prow/go/src/github.com/kubernetes-sigs/node-feature-discovery/pkg/utils/memory_resources.go:54:17: n.Name undefined (type fs.DirEntry has no field or method Name) 
 ```
 should I rebase my branch to fix it? @marquiz

**Comment by marquiz** [comment]
> should I rebase my branch to fix it? @marquiz

Hmm, no, the logcheck seems to be flaky 🧐
/retest

**Pull request merged**
